### PR TITLE
Fixes #2233: display registered providers

### LINKED ISSUE #2233: Add display of registered providers to FakedWBEMConnection display_repository
We removed the method callback and removed it from the display_repository.  However, we never added a method to display currently registered providers.  I suggest we add that to display_repository.

### PR DESCRIPTION
**Pull request by KSchopmeyer**
Complete.  Adds display of registered provider and tests.

In latest revision I made the _provider_namespace, etc. public methods.

### CODE REVIEW

File: pywbem_mock/_wbemconnection_mock.py [hunk 1]
```diff
@@ -128,6 +134,51 @@ def __repr__(self):
...
+        def print_ljust(rows):
+            """
+            Print left justified and column aligned row of rows where each item
+            is either a string or list of strings
+            """
+            # Convert lists of strings to joined string
+            for i, row in enumerate(rows):
+                for j, item in enumerate(row):
+                    if isinstance(item, list):
+                        rows[i][j] = "[{0}]".format(", ".join(item))
```
Comment (andy-maier) [inline pywbem_mock/_wbemconnection_mock.py:163]: Can there ever be a list of strings?

File: pywbem_mock/_wbemconnection_mock.py [hunk 2]
```diff
@@ -128,6 +134,51 @@ def __repr__(self):
...
+        Generate display of registry in readable form and return the string
+        Output format is:
+        <namespace>:
+          <classname>: CIM_Namespace provider: NamespaceProvider type: instance
+
+        For example:
+        root/cimv2:
+          CIM_Namespace, provider: NamespaceProvider type: instance
+          CIM_Blah, provider: BlahProvider, type: instance
+
```
Comment (andy-maier) [inline pywbem_mock/_wbemconnection_mock.py:148]: Reading the code (I did not try it out), the output is for example:
 `CIM_Blah  BlahProvider  instance`
Ie not including keywords like provider or type.

File: pywbem_mock/_wbemconnection_mock.py [hunk 3]
```diff
@@ -128,6 +134,51 @@ def __repr__(self):
...
+        widths = [max(map(len, col)) for col in zip(*rows)]
+        for row in rows:
+        print "  ".join((val.ljust(width) for val, width in zip(row, widths)))
```
[the remark this answers is not in the document]
Comment (KSchopmeyer) [reply pywbem_mock/_wbemconnection_mock.py:151]: Sorry. A leftover as I was modifying code.

File: pywbem_mock/_wbemconnection_mock.py [hunk 4]
```diff
@@ -128,6 +134,51 @@ def __repr__(self):
...
             s=self)
 
+    def display_registered_providers(self, dest=None):
+        """
+        Generate display of registry in readable form and return the string
+        Output format is:
+        <namespace>:
+          <classname>: CIM_Namespace provider: NamespaceProvider type: instance
+
+        For example:
```
[the remark this answers is not in the document]
Comment (KSchopmeyer) [reply pywbem_mock/_wbemconnection_mock.py:144]: Done.  Added to both conn method and provider_registry method

File: pywbem_mock/_wbemconnection_mock.py (hunk 1 above)
Comment (KSchopmeyer) [reply pywbem_mock/_wbemconnection_mock.py:163]: Actually turns out no so removed.

File: pywbem_mock/_wbemconnection_mock.py (hunk 2 above)
Comment (KSchopmeyer) [reply pywbem_mock/_wbemconnection_mock.py:148]: I modified the comment to match the current code. Fixed example also.

File: pywbem_mock/_wbemconnection_mock.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (KSchopmeyer) [reply pywbem_mock/_wbemconnection_mock.py]: DONE

**Comment by andy-maier** [comment]
The only change left is the output format. See discussion in the Slack channel.

DISCUSSION Do we want to do that in this PR or separately?. My take would be this PR.

**Comment by andy-maier** [comment]
On iterating over the registered providers: Discussion in the Slack channel suggests to simply make the internal methods public, in a separate PR.

DISCUSSION

**Comment by andy-maier** [comment]
The update you mentioned in the Slack channel is not yet up. Pls double check.

File: pywbem_mock/_wbemconnection_mock.py [hunk 5]
```diff
@@ -348,26 +401,103 @@ def get_registered_provider(self, namespace, provider_type, classname):
...
+
+    def provider_namespaces(self):
+        """
+        Get iterable of namespaces for registered providers
+
+        Returns:
+            case-insensitive :term:`py:iterable` of :term:`string`: namespaces
+            for which providers are registered.
+        """
+        return self._registry
```
[the remark this answers is not in the document]
Comment (KSchopmeyer) [reply pywbem_mock/_wbemconnection_mock.py:433]: Actually the documentation says it returns a case-insensitive iterable

Changed just o get this thing committed.

File: pywbem_mock/_wbemconnection_mock.py [hunk 6]
```diff
@@ -348,26 +401,103 @@ def get_registered_provider(self, namespace, provider_type, classname):
...
+
+        Returns:
+            case-insensitive :term:`py:iterable` of :term:`string`:
+            Names of classes registered for namespace
+
+        Raises:
+          KeyError: if namespace invalid
+        """
+
+        return self._registry[namespace]
```
[the remark this answers is not in the document]
Comment (KSchopmeyer) [reply pywbem_mock/_wbemconnection_mock.py:453]: Changed.

**Pull request merged**